release-22.1: sql: block operations properly on implicit transactions

fqazi opened the pull request:
Backport 2/2 commits from #78512.

/cc @cockroachdb/release

---

Previously, an enhancement was made to allow implicit
transactions to execute multiple statements, which caused
code to block operations in implicit transactions to break.
This was problematic because these scenarios cared that
only a single statement is executed, and not if it was
only an implicit transition. For example, the declarative
schema changer and legacy schema changer could be mixed
in implicit transactions after this change. To address this,
this patch modifies places that we're checking for implicit
transactions to instead check for single statement transactions.

Release note: None
Release justification: Low risk and mainly focused on fixing behaviour with a disabled feature for multiple statements in implicit transactions


Jira issue: CRDB-14719

irfansharif committed:
spanconfig: squash benign data race

Fixes #75849.

Release note: None

irfansharif committed:
spanconfig: verify migration for rangefeed enablement

\#74555 starts using the span configs infrastructure to control whether
rangefeeds are enabled over a given range. Before dynamic system table
IDs (#76003), we used the range's key boundaries to determine whether
the range in question was for a system table ID. In mixed-version
clusters, it's possible to have both forms of this check. To ensure
things work in this form (something we suspected in #76331), we add a
test.

NB: The reason things still work is because in #74555 we modified the
system config span to hard code the relevant config fields for constant
system table IDs -- behaving identically to previous version nodes.

Release note: None

tbg committed:
roachtest: remove extraneous EncryptDefault(false)

The methods that actually start nodes in this test already specify
`DontEncrypt`[^1][^2].

Release note: None

[^1]: https://github.com/cockroachdb/cockroach/blob/21a286f2b2978b1cc117596865cd680dd268cf74/pkg/cmd/roachtest/tests/versionupgrade.go#L373-L374
[^2]: https://github.com/cockroachdb/cockroach/blob/21a286f2b2978b1cc117596865cd680dd268cf74/pkg/cmd/roachtest/tests/versionupgrade.go#L422-L423

tbg committed:
roachtest: avoid Encrypt{Default,AtRandom} in tpccbench

Use the `DontEncrypt` bool for `StartOpts` instead.

Release note: None

tbg committed:
roachtest: simplify randomized encryption.

It was easy to hold on. @AlexTalks noticed in
https://github.com/cockroachdb/cockroach/pull/76568 that even for a test
that opted out of randomized encryption, it could end up with a cluster
that was using encryption. I also generally found the interplay between
`EncryptDefault` and `EncryptAtRandom` hard to understand.

It simplified this now:

- EncryptDefault is gone
- EncryptAtRandom does what you think (it did that before)
- a reused cluster behaves like a new cluster.

There is still a bit of "weird stuff" which I attempted to clean up.
Encryption can be enabled in one of two ways.

You can either run `roachtest run --encrypt` which enables encryption
for all clusters, so the randomization doesn't matter. Without the
encrypt flag, randomization only applies to tests that *opt in*, via
`EncryptAtRandom`. The question is what tests like `tpccbench` want to
do; they currently hard-code encryption to off (i.e. they'll ignore the
`--encrypt` flag by setting `DontEncrypt` on their start opts) which
doesn't seem quite right; if I want to run tpccbench with `--encrypt`
that's what I should be getting. I therefore removed these workarounds
for all tests that employed them; if we run tests in nightlies with
`--encrypt` (which we don't anyway) then the right solution would
be for tests to detect that and to skip themselves. We can do that
when the need arises.

Release note: None

samiskin committed:
jobs: fix startable job num-runs overcount

Release justification: low risk bug fix

Removes extra call to started() in StartableJob's Start since its
already called in stepThroughStateMachine later on

Release note (bug fix): fix num_runs being incremented twice for certain
jobs upon being started

sjbarag committed:
cluster-ui: upgrade jest and related dependencies

Jest before v27 didn't support configuration options that allow test
files to be discovered across symlink boundaries, which is necessary
when running Jest tests within a Bazel sandbox. Upgrade Jest and its
dependencies (including a manual version resolution of
jest-environment-jsdom [1] defined in pkg/ui/package.json to avoid Jest
bugs [2]) to enable cluster-ui tests being run via Bazel.

[1] https://github.com/enzymejs/enzyme-matchers/issues/351#issuecomment-976009553
[2] https://github.com/yarnpkg/yarn/issues/5039

Release justification: Upgrade of test dependencies
Release note: None

otan committed:
builtins: add crdb_internal.validate_ttl_scheduled_jobs

Release justification: high benefit addition to new functionality

Release note (sql change): Added a
`crdb_internal.validate_ttl_scheduled_jobs` builtin which verifies each
table points to a valid scheduled job which will action the deletion of
expired rows.

otan committed:
descs: introduce GetTableNameByID / GetTableNameByDesc

This commit introduces the GetTableNameByDesc and GetTableNameByID
functions, which fetches a *tree.TableName from the given table objects.

Release justification: low-risk refactor

Release note: None

otan committed:
jobs: pass descs.Collection to ShowCreateStatement

In preparation for the showing table names for scheduled jobs in SHOW
CREATE SCHEDULE.

Release justification: low-risk update for new functionality

Release note: None

otan committed:
jobs: add descsCol to OnDrop method

In preparation for the next commit.

Release justification: low-risk changes for new functionality

Release note: None

otan committed:
ttlschedule: allow dropping invalid schedules

Previously, `DROP SCHEDULE` for a TTL job would fail always. Now, we
succeed if the scheduled job is invalid.

Release justification: low risk high benefit change to new functionality

Release note: None

msbutler committed:
backupccl: allow SHOW BACKUP FILES/RANGES/SCHEMAS in backup collections

Fixes #77697

Release justification: low risk bug fixes

Release note: None

sjbarag committed:
bazel: link internal dependencies in pkg/ui/ monorepo

During the `yarn install` phase of a yarn workspace-based monorepo (like
the one rooted in ./pkg/ui), yarn sets up symlinks in the root
node_modules/ directory that point at the packages included within that
workspace. This allows standard nodejs module resolution to find
external and monorepo-internal dependencies to be found transparently
with no additional tool configuration. Unfortunately, Bazel's
`rules_nodejs` doesn't do that by default. Ensure intra-project
dependencies are properly linked so they can be found during test runs
via Bazel.

Release justification:
Release note: None

sjbarag committed:
db-console: use BAZEL_TARGET environment variable to detect Bazel builds

The db-console webpack configuration previously relied on a custom
variable (passed via webpack's `--env` CLI flag) to detect builds
triggered from Bazel. Luckily Bazel already sets the `BAZEL_TARGET`
environment variable for all commands it runs, so our custom
`--env.is_bazel_build` CLI argument isn't necessary. Switch to
the BAZEL_TARGET environment variable for Bazel build detection.

Release justification: Build system change
Release note: None

sjbarag committed:
db-console: move polyfill out of src/js

The tree at pkg/ui/workspaces/db-console/src/js was used almost
exclusively for the @cockroachlabs/crdb-protobuf-client package, with
the exception of one polyfill that's used only during tests. Move the
Object.assign polyfill to .../src/util/polyfills/ so that the
.../src/js/ tree can be completely ignored when not building the
protobuf client.

Release justification: Internal refactor to simplify tests
Release note: None

sjbarag committed:
cluster-ui: don't run eslint through Jest

esbuild -- a common linter for JavaScript and JS-related projects -- is
currently being run two different ways: via its default CLI and via the
Jest test runner as a separate "project" of tests. Having multiple
Jest projects (one for actual tests and one for lints) introduces a bit
of extra complexity that isn't strictly necessary, and makes
conversations about linting ambiguous. Remove the "linting" project from
Jest, so that eslint is only executed via its default CLI.

Release justification: Build system change
Release note: None

sjbarag committed:
dev: add silent-mode to OS command abstractions

The OS abstraction layers used widely within `./dev` would previously
log every `cp` and `mkdir` command used. This behavior complicates
testing in some cases, particularly when copying entire output trees.
Allow `./dev` commands to selectively silence and un-silence the echoing
of commands that are run to make testing easier.

Release justification: Build system change
Release note: None

sjbarag committed:
dev: add os.RemoveAll and os.CopyAll functions

It's sometimes necessary to recursively copy a directory or recursively
delete a non-empty directory, but the OS abstractions used by `./dev`
didn't previously offer that functionality. Add equivalents of `rm -rf`
and `cp -r` to the `os` module used by `./dev`.

Release justification: Build system change
Release note: None

sjbarag committed:
dev: implement ./dev ui test with watch mode

Running unit tests for the CRDB UI subprojects previously required the
use of the `make ui-test` (and `make ui-test-watch`) targets. Allow the
Makefile to (eventually) be removed by offering a pair of equivalent
commands using the `./dev` abstraction:

* make ui-test       -> ./dev ui test
* make ui-test-watch -> ./dev ui test --watch

Release justification: Build system change
Release note: None

sjbarag committed:
dev: use Bazel-provided version of yarn CLI

While UI devs likely have `yarn` installed and on their `$PATH`, the
previous assumption of that fact isn't necessary: the Bazel build system
provides its own `yarn` CLI. Use the automatically-downloaded copy of
`yarn` instead of the globally-installed one, to ensure consistency
between users.

Release justification: Build system changes
Release note: None

otan committed:
rfcs: fix typo in RFC

Release justification: non-prod code change
Release note: None

sjbarag committed:
db-console,cluster-ui: run linters through Bazel

UI sub-projects were traditionally linted via the `make ui-lint` target
but the Make-based build is being deprecated soon. Specifically, that
make target would run:

* In pkg/ui/workspaces/db-console:
  * stylint (linting for CSS and several "compiles to CSS" dialects)
  * eslint (linting for JavaScript and TypeScript)
  * prettier (through eslint; formatting for JavaScript and TypeScript)
  * tsc (the TypeScript compiler, for type-checking only)
* In pkg/ui/workspaces/cluster-ui:
  * eslint (linting for JavaScript and TypeScript)
  * prettier (through eslint; formatting for JavaScript and TypeScript)

The type-checking pass for db-console is currently broken and isn't
ported to Bazel, but a type-checking Bazel target is now present for
cluster-ui where one didn't exist under make. Allow static analyzers for
pkg/ui/ projects to be run through Bazel, so the Makefile can be
(eventually) removed.

Release justification: Build system changes
Release note: None

sjbarag committed:
ui: add ui-root-level lint target

Linting targets for db-console and cluster-ui were previously
un-grouped, requiring a dev to know that they exist and call `bazel
test` for both targets together. Add a single test suite that triggers
all UI-related static checks.

Release justification: Build system change
Release note: None

sjbarag committed:
dev: add ./dev ui lint subcommand

Release justification: Build system change
Release note: None

stevendanna committed:
sql: track index merging completion in mvcc-compatible backfiller

This adds a naive basic fraction completion calculation to the merge
phase of the index backfill. The completion fraction of the backfill
and merge are scaled by an arbitrarily chosen guess at what portion of
the total process they account for.

Release justification: Bug fix for newly added feature.
Release note: None

sjbarag committed:
dev: upgrade to v23

Upgrade dev to v23 to ensure users get the new `./dev ui lint`
subcommand.

Release justification: Build system change
Release note: None

rickystewart committed:
bazel: use starlark flags for these configurations instead of `define`s

I thought this might result in some build perf improvements but haven't
noticed any in my testing. Having said that, this is still an
ergonomic improvement.

Also add aliases so you can do e.g. `--dev` instead of `--config dev`.

Release note: None

tbg committed:
roachtest: move encryption randomization to a better place

Release note: None

tbg committed:
roachtest: phase out DontEncrypt arg

A number of tests were previously opting out of encryption-at-rest using
the `DontEncrypt` option in `c.Start`. I suspect this dates back to a
time in which we were erroneously re-using encrypted clusters for
subsequent tests, i.e. we had tests that hadn't explicitly opted into
encryption at rest hand encrypted clusters to them anyway, which not
all of them supported (in particular, some tests perform raw
manipulation of the CRDB data).

Now, only tests that opt into encryption at rest (via their `TestSpec`
can accept encrypted clusters).

A behavior change occurs when running roachtest with `--encrypt`:
Previously the tests that explicitly opted out would use a non-encrypted
cluster anyway. But this was confusing behavior.

We're not running roachtest nightly with forced encryption for all
tests, so we take the stance that all tests must support unencrypted;
they may support encrypted but invoking a test that does not support
encryption with the `--encrypt` flag will simply fail. (If we desire to
run "all tests that support it" under encryption in the future, we can
bring an opt-out back; it looks like many of the opt-outs removed here
were no longer necessary, so I don't feel bad ripping this out even if
there's a chance that we'll bring it back later).

Release note: None

tbg committed:
roachtest: add test verifying effect of EncryptAtRandom

This verifies that opting into EncryptAtRandom via the test spec
probabilistically results in a cluster that uses encryption at
rest.

Release note: None

tbg committed:
roachtest: respect --encrypt=random

As of the previous commit, when randomizing encryption-at-rest, we were
not taking into account the `--encrypt` flag. We were only supposed to
randomize encryption (with either 50% or 100% probability) when
`--encrypt` was set to `random` or `true`.

I am not sure how this all worked prior to the PR containing this
commit. We are not passing the `--encrypt=random` flag to our nightly
roachtests (but we are passing it to the weekly ones from the looks
of it). I think that because tests explicitly set the cluster to
randomly encrypt, the flag was in effect ignored for those cases.

It's also not too important what the old behavior was, what matters is
that we get sane behavior now:

- when `--encrypt` not specified, nobody encrypts.
- at `random`, only tests that support random encryption randomly
  encrypt.
- at `true`, all tests get an encrypted cluster, whether they like it or
  not.
- nightly roachtest now explicitly specifies --encrypt=random.

Adding a test in the `roachtest` package also forced me to fix up these
tests under bazel (since the bazel stressracer would still pick up the
new test, even though the package as a whole was excluded under bazel).
So we get to remove the broken-under-bazel tag. This required only some
mild mocking.

Release note: None

Commit:
sql: apply zone configs to temp indexes during backfill

Previously applying a zone config to a new index that's backfilling only
applied the config to the new index, and not the corresponding temporary index
created for the backfill. Change this so that the temporary indexes also get
the same zone configs as the new index.

Release justification: fixes temp indexes not getting zone configs
Release note: None

adityamaru committed:
backupccl: fix paper cut in latest files directory structure

Prior to this change, the LATEST files of a backup were written to
`metadatalatest/` instead of `metadata/latest/`, this patch fixes
that.

Release note (bug fix): The LATEST file that points to the latest full
backup in a collection was written to a directory path with the wrong
structure.

Commit:
ui: Hide table stats collection setting for non-admins

Release justification: low-risk change

Release note: None

mgartner committed:
tree: correctly handle NULLs annotated as a tuple type

Fixes #74729

Release note (bug fix): A bug has been fixed that caused errors when
trying to evaluate queries with NULL values annotated as a tuple type,
such as `NULL:::RECORD`. This bug was present since version 19.1.

Santamaura committed:
ui: handle latency not defined on network page

Previously when a cluster with multiple nodes had a node stopped
and then another node quickly started, the network page would crash.
This was due to the node object holding the latency value being
undefined and the ui trying to read this key when that was undefined.
This occurs when a node is in an `UNAVAILABLE` state.
This patch resolves the issue by being more defensive on the front
end by safely attempting to access latency and if it is undefined,
set the value to 0. The existing code is able to handle this case
afterward and will eventually set the user friendly latency status
to `--`.
Resolves: #59322

Release note (ui change): Fixes a bug where a node in the
`UNAVAILABLE` state will not have latency defined and cause the
network page to crash.

nicktrav committed:
roachtest: sstable-corruption: collect manifests; additional logging

Add additional logging of the contents of the data directory and collect
the Pebble manifests when the test fails to find tables to corrupt. This
will aid in diagnosing the test failures.

Touches #78121.

Release note: None.

otan committed:
sql: implement crdb_internal.repair_ttl_table_scheduled_job

This commit is intended for "repairing" jobs that may have been broken
by a broken schema change interaction.

Release justification: high priority fix for new functionality

Release note (sql change): Adds a
`crdb_internal.repair_ttl_table_scheduled_job` builtin, which repairs
the given TTL table's scheduled job by supplanting it with a valid
schedule.

otan committed:
sql: amend COPY to use CRDB csv implementation

Release note: None

rharding6373 committed:
sql: fix migration with new system.table_statistics column

Before this change, the new `system.table_statistics` column `avgSize`
introduced in version 21.2.12 was appended to the end of the table
during migration, but the system schema had the new column in a
different order. The column was also not added to the existing column
family containing all table columns during migration.

This change fixes both the system schema and the migration commands so
that the column ordering is the same and the new column is added to the
existing column family. Unfortunately, this means that the existing
column family name is unable to be updated to include the column.

Fixes: #77979

Release justification: Fixes a schema migration bug in the
table_statistics table.

Release note: None

otan committed:
parser: add unimplemented syntax for all the COPY options

Release note: None

otan committed:
tree: add pgcodes to multiple specified COPY options

Release note: None

xinhaoz committed:
sql: record index usage stats in index joins

Fixes #76173

Previously, the construction of index joins skipped the
recording of the  primary key for index usage stats.
This commit records the use of the primary key if the
statement is not of the EXPLAIN variety.

Release note (bug fix): Index usage stats are now properly
captured for index joins.

otan committed:
ttlschedule: display table name for SHOW CREATE SCHEDULE for TTL

Release justification: low-risk update to new functionality

Release note: None

mgartner committed:
internal/sqlsmith: fix join expr generation

Join generation logic in sqlsmith assumed that all equivalent types can
be compared, except for `void`. However, there are other types that
cannot be compared, like `tuple[]`. This commit ensures that sqlsmith
only generates joins with scalar types to avoid panics.

Fixes #78149

Release note: None

Azhng committed:
sql: new SQL Stats cluster settings to improve write traffic

Resolves #78339

Previously, SQL Stats flushed to system table as soon as the in-memory
buffer is full. This means  the size of the system tables that back SQL
Stats could grow faster than the cleanup job. Additionally, when the
SQL Stats flush is disabled, the SQL Stats is unable to collect any more
new statement / transaction statistics when the in-memory store is full.

This commit introduces two non-public cluster settings:
* `sql.stats.flush.minimum_interval`: this setting limits minimum interval
  between each flush operation. If a flush operation is triggered sooner
  than what is allowed by the minimum interval, (e.g. when the in-memory
  SQL Stats store is full), the flush operation is aborted. By default
  this cluster setting is set to 0.
* `sql.stats.flush.force_cleanup.enabled`: which allows the
  in-memory SQL Stats to be cleared at the interval specified by
  `sql.stats.flush.interval`, even if the SQL Stats flush is disabled.
  By default, this cluster setting is set to false.

This commit also updated the stmt_grouping_in_explicit_txn data driven
test to ensure the output order is deterministic.

Release note: None

otan committed:
encoding/csv: add escape logic

Add `ESCAPE` logic to the `encoding/csv` package, for exposure to
SQL at a later stage.

It is worth noting I wrote this in the "safest" backportable way
possible. Ideally we'd rewrite the read logic to be more "parser"-like
to account for the change in QUOTE case, but that's a lot riskier to
backport.

Release note: None

otan committed:
sql: implement COPY FROM ... ESCAPE ...

Release note (sql change): Implemented the `COPY FROM ... ESCAPE ...`
syntax.

msbutler committed:
backupccl: add FROM keyword to SHOW BACKUP

Previously, users could run SHOW BACKUP on collections using:
`SHOW BACKUP {subdir} IN {DEST}`
This commit allows the user to run an identical SHOW BACKUP via:
`SHOW BACKUP FROM {subdir} IN {DEST}`
Adding the `FROM` better aligns the SHOW BACKUP syntax with RESTORE.

All non-default show backup commands (e.g. SHOW BACKUP SCHEMAS/RANGES/FILES)
must be run with FROM. This UX change does not affect any released binaries as
these non-default SHOW BACKUP commands were not available on collections until

This commit also refactors the grammar for SHOW BACKUP in sql.y to make future
 refactors to SHOW BACKUP syntax a bit easier.

Release note (sql change): Allow users to run SHOW BACKUP on collections with
the FROM keyword: `SHOW BACKUP FROM {subdir} IN {DEST}`. Users can still run
SHOW BACKUP without the FROM keyword; however, for 22.1 docs, we should only
document SHOW BACKUP with the FROM keyword, once the backport merges. One
gotcha: `SHOW BACKUP SCHEMAS` must be run with FROM, though this shouldn't
affect our documentation strategy for this commit.

fqazi committed:
sql/logictest: re-enable declarative schema changer for enums test

Fixes: #78029

Previously, the enums logictest could potentially hit an
infinite number of retries inside with the declarative
schema changer enabled. As a result when enabling the
declarative schema changer we temporarily modified the
enums test to explicit disable the declarative schema
changer. To address this, this patch will re-enable the
declarative schema changer for the enums test.

As a part of this work, we noticed that enhancements to
implicit transaction behaviour to batch changes in a single
transaction also exposed interactions between the declarative
and legacy schema changer. To avoid issues these statements
were intentionally split inside the logicttest.

Release note: None

arulajmani committed:
roachpb,spanconfig: add GetAllSystemSpanConfigsThatApply RPC

This patch adds the `GetAllSystemSpanConfigsThatApply` RPC to the
`Connector` and has the `spanconfig.KVAccessor` implement it. This RPC
returns all system span configs that apply to a given tenant. This
is only ever meaninful in the case of secondary tenants. It serves as a
means for secondary tenants to get a view into the system span config
state that applies over their ranges set by the system tenant.

We'll make use of this RPC in an upcoming patch. We'll use it to make
GC decisions when GC-ing dropped tables/indexes in the GC job.

Release note: None

otan committed:
encoding/csv: fix comment typo for Escape rune

Release note: None

fqazi committed:
sql: block operations properly on implicit transactions

Previously, an enhancement was made to allow implicit
transactions to execute multiple statements, which caused
code to block operations in implicit transactions to break.
This was problematic because these scenarios cared that
only a single statement is executed, and not if it was
only an implicit transition. For example, the declarative
schema changer and legacy schema changer could be mixed
in implicit transactions after this change. To address this,
this patch modifies places that we're checking for implicit
transactions to instead check for single statement transactions.

Release note: None

yuzefovich committed:
execinfrapb: break the dependency on sem/builtins

This commit moves several utility functions from `execinfrapb` into
`execinfra` package in order to break the dependency of the former on
`sem/builtins` (which eventually depends on the `c-deps`).

Release note: None

arulajmani committed:
gcjob: teach GC job to respect protected timestamps for tables/indexes

The GC job is responsible for issuing clear ranges for dropped tables
and indexes. Previously, we weren't consulting the new PTS subsystem
when deciding if a table or an index's range could be cleared. This
patch fixes that.

We check if a table/index span is protected after the GC TTL expires.
We do so by issuing an RPC to KV to fetch all span configurations and
system span configurations that apply over the spans. For secondary
tenants, this also includes any system span configurations that the
host tenant may have set over its ranges.

Lastly, it's worth noting that we make no attempt to coordinate with
concurrent protected timestamp writers. This is okay, given that
protected timestamps make no claims about expired data.

Closes #77156

Release note: None

rickystewart committed:
cluster: Revert "cluster: use WaitConditionNextExit"

This reverts commit 65437495b20caf547e9354ff05e433b9067e7624.
That commit was an (unsuccessful) attempt to fix #58955, and in the
presence of this change the `acceptance` tests are very likely to hang
forever under Ubuntu 20.04 due to a race condition where the container
exits before we begin waiting on it.

Release note: None

Commit:
sql/catalog/lease: handle ambiguous errors

Before this patch, ambiguous errors from kv during lease acquisitions would
both make their way to the client and would result in a leaked lease. This
leaked lease caused a flakey versionupgrade test.

Along the way I needed to make `kv.AutoCommitError` behave correctly in terms
of error wrapping.

Fixes #73719.

Release note (bug fix): Fixed a bug whereby certain catalog interactions which
occurred concurrently with node failures were not internally retried.

rickystewart committed:
bazel: fix `sql-gen.sh` script

This command was cargo culted from the `Makefile`, which uses double
`$$` because the `$` needs to be escaped in the `make` context. In the
`bash` context that this script runs in, however, `"$$ret"` evalutes to
the value of the special variable `$$` (the PID of the current process)
concatenated to the string `ret`. Obviously, this string will never
contain the word `"conflicts"` so this means the Bazel build is unable
to detect the error case where the SQL grammar has conflicts.

Release note: None

nicktrav committed:
roachtest: de-flake sstable corruption test

Currently, the `sstable-corruption/table` randomly selects a number of
SSTables on each node to corrupt, before restarting the cluster and
running a workload in an attempt to notice the corruption. The test uses
a monitor to monitor for node death, and also run the background
workload.

There exists a race where monitor's workload goroutine can exit soon
after starting (e.g. due to #77627), and _before_ the monitor encounters
the failure of a node. The test expects an error to be returned by the
monitor, which fails due to the former error being nil.

Address the race by continuing to run the workload even in the case of
failure, exiting only when a timeout is exceeded (due to not
encountering corruption within the allotted time), or other context
cancellation (i.e. a node death _was_ observed by the monitor and the
test is being torn down).

Release note: None.

rickystewart committed:
ci: add scripts for bazel-based `ui test`, `ui lint` ci jobs

These scripts just do the same logic that `dev` does for the same
functions.

Release note: None

sjbarag committed:
ui: don't download .dll.js bundles

Loading the CockroachDB admin UI (served from port 8080) would
previously download three JS files: `bundle.js`, `vendor.dll.js`, and
`protos.dll.js`. Neither `vendor.dll.js` nor `protos.dll.js` need to be
included via `<script/>` tag, as they're used only as build-time tooling
to accelerate webpack builds. Only a single file -- `bundle.js`
-- needs to be loaded for the admin UI to be functional, and
Bazel-produced builds don't have any `.dll.js` files to serve. Don't
load `.dll.js` files via index.html.

Release note (performance improvement): Unused JS files are no longer
downloaded when the admin UI loads.

AlexTalks committed:
clusterversion: introduce version key for `crdb_internal.cluster_locks`

This change adds a new cluster version to enable usage of the virtual
table `crdb_internal.cluster_locks`, so that we can disable usage of new
KV `QueryLocksRequest` API on cluster versions prior to 22.1.

Release note: None

Release justification: introduces a new cluster version without any
logic changes

sjbarag committed:
build: don't include .dll files in binary

Since dll file aren't being served directly to users anymore, there's no
need to embed them in the `cockroach` binary. Stop copying `.dll.js`
files into `pkg/ui/dist{oss,ccl}` during builds, reducing the
`cockroach` binary size by ~19MB (darwin, x86) when built with GNU make.

Release note: None

dt committed:
jobs: return the actual error that requested a pause

Explain that the job was paused, why, and what to do about it.

Release note: none.

dt committed:
sql: add pause reason to SHOW JOBs

If a job is paused with reason, show that reason in the running_status field.

Release note: none.

dt committed:
ui/db-console/views/jobs: use WithErrorMessage if error is set

Paused jobs can be paused because of an error; if an error is set, even if the status is paused
rather than failed, show the error message.

Release note: none.

yuzefovich committed:
colexecproj: unset nulls in the default comparison operator

Previously, we forgot to unset nulls in the default comparison operator
(which handles all comparisons that are not natively supported by the
vectorized engine), so whenever an output vector is reused, we could
produce an incorrect output. However, in most cases we got lucky because
`ResetInternalBatch` unsets nulls on all vectors, even if those aren't
owned by the caller of the method. (For context, our projection
operators append its output vector to a batch that is otherwise owned by
another operator, like `ColBatchScan`.) Thus, I decided to not include
a release note since the bug appears to be quite rare. There is also no
regression test since I'm working on improving the test harness that
would have caught this (currently, the test harness itself calls
`ResetInternalBatch` which hides this bug in the unit tests).

Release note: None

rharding6373 committed:
sqlsmith: fix deadlock during generation

Fixes a deadlock when sqlsmith generates join expressions.

Fixes: #78555

Release note: None

ajwerner committed:
sql: drop locks before waiting for concurrent schema changes

Before this change, it was possible for a deadlock to occur when
one schema change was waiting on a previously comitted schema change
but the job due to that previously committed schema change was
waiting on the current transaction. It is critical that when waiting
for a concurrent schema change, that the transaction doing the waiting
not hold any locks.

Along the way, this change reworks how "old" schema changes interact
with the declarative schema changer. Before this change, they would return
an error. Now they'll wait. I don't think the error that we were returning was
going to be a good experience for anybody, so this seems like a strictly good
change.

Lastly, this change ensures that when explicit transactions encounter such
waiting, that they return the proper error code.

Fixes #78144

Release note: None

jaylim-crl committed:
ccl/sqlproxyccl: add connection tracker component

This commit adds a connection tracker component to the balancer, and this will
be used to track all instances of forwarders for each tenant. The forwarders
can then be used for connection migration.

Release note: None

tbg committed:
roachpb,kvserver: tweak ReplicaUnavailableError rendering

Release justification: minor UX tweakto an error message
Release note: None

rytaft committed:
opt: disable index recommendations with PARTITION ALL BY

This commit disables index recommendations for tables with
PARTITION ALL BY (including REGIONAL BY ROW tables) because the
current recommendations are not valid. This is a backportable
fix, but a future PR will fix these recommendations and reenable
them.

Release note (sql change): Disabled index recommendations in EXPLAIN
output for REGIONAL BY ROW tables, as the previous recommendations were
not valid.

msbutler committed:
backupccl: add deprecation warning for SHOW BACKUP without subdir

Informs: #78153

Release note (sql change): The `SHOW BACKUP` cmd  without the `IN` keyword to
specify a subdirectory is deprecated and will be removed from a future release.
Users are recommended to only create collection based backups and view them
with `SHOW BACKUP FROM <backup> IN <collection>`

rickystewart committed:
ci: make sure we're streaming test output when `stress`ing via Bazel

Without this, you don't see the helpful "ran X iterations so far, Y
failures" messages from `stress`.

Release note: None

nvb committed:
sql/rowexec: deflake TestUncertaintyErrorIsReturned

Fixes #78610.

This commit deflakes `TestUncertaintyErrorIsReturned` by scoping
its filter to only KV Scans against the SQL table that the test
is querying.

I'm not entirely sure why this become flaky after 208e2b4, but
what follows is my best theory:
- the test uses replication factor 1
- the test performs a series `ALTER TABLE t EXPERIMENTAL_RELOCATE` ops
- before 208e2b4, these operations we handled as a multi-step process
  which involved adding a new voter, moving the lease, demoting the
  original voter, and removing the original voter.
- after 208e2b4, these operations all occur in a single joint-configuration.
- (conjecture) this compressed replica movement was able to complete without
  the client needing to update its range cache. As a result, the test's query
  needed to refresh the range cache after the filter was installed. However,
  because the filter was rejecting all traffic, it was rejecting even the
  meta range lookups (I saw this). So the test got stuck spinning on meta
  lookups.

Regardless of the exact explanation, this is a test flake and not an indication
of anything more serious.

Release justification: fixes test flake.

Commit:
tabledesc: stop promoting encoding version counters for secondary idxs

This commit fixes a bug which prevents the correct operation of
a mixed-version cluster: when a table is created with a secondary index
on a 22.1 node, its descriptor will fail validation on a 21.2 node.

To prevent this, this commit no longer sets the version of
secondary indexes to the latest value, but to the value below it.
This is functionally exactly identical for secondary indexes, the latest
value only changes expectations for primary indexes.

Fixes #78426.

Release justification: correctness bug fix
Release note: None

RichardJCai committed:
sql: support SHOW DEFAULT PRIVILEGES IN SCHEMA

Release note (sql change): Add support for show default privileges
in schema.

The SHOW DEFAULT PRIVILEGES clause now supports optionally passing a
schema name.

SHOW DEFAULT PRIVILEGES [opt_for_role] [opt_schema_name]

Example:

SHOW DEFAULT PRIVILEGES IN SCHEMA s2
----
role      for_all_roles  object_type  grantee    privilege_type
testuser  false          tables       testuser2  DROP
testuser  false          tables       testuser2  SELECT
testuser  false          tables       testuser2  UPDATE

SHOW DEFAULT PRIVILEGES FOR ROLE testuser IN SCHEMA s2
----
role      for_all_roles  object_type  grantee    privilege_type
testuser  false          tables       testuser2  DROP
testuser  false          tables       testuser2  SELECT
testuser  false          tables       testuser2  UPDATE

mgartner committed:
opt: do not cross-join input of semi-join

This commit fixes a logical correctness bug caused when
`GenerateLookupJoins` cross-joins the input of a semi-join with a set of
constant values to constrain the prefix columns of the lookup index. The
cross-join is an invalid transformation because it increases the size of
the join's input and can increase the size of the join's output.

We already avoid these cross-joins for left and anti-joins (see #59646).
When addressing those cases, the semi-join case was incorrectly assumed
to be safe.

Fixes #78681

Release note (bug fix): A bug has been fixed which caused the optimizer
to generate invalid query plans which could result in incorrect query
results. The bug, which has been present since version 21.1.0, can
appear if all of the following conditions are true: 1) the query
contains a semi-join, such as queries in the form:
`SELECT * FROM t1 WHERE EXISTS (SELECT * FROM t2 WHERE t1.a = t2.a);`,
2) the inner table has an index containing the equality column, like
`t2.a` in the example query, 3) the index contains one or more
columns that prefix the equality column, and 4) the prefix columns are
`NOT NULL` and are constrained to a set of constant values via a `CHECK`
constraint or an `IN` condition in the filter.

jaylim-crl committed:
ccl/sqlproxyccl: make forwarder implement balancer.ConnectionHandle

This commit makes the forwarder struct implement the balancer.ConnectionHandle
interface. This then allows the forwarder to be stored within the connection
tracker component.

Release note: None

jaylim-crl committed:
ccl/sqlproxyccl: register forwarders with the connection tracker

This commit registers all forwarders with the connection tracker. At the same
time, we remove the afterForward testing hook since that is no longer needed.

Release note: None

yuzefovich committed:
colexecop: reset the internal batch in a test operator

This commit makes a test operator that "owns" a batch to properly reset
the batch on each `Next` call. This is the expectation that we forgot to
follow.

This was discovered when running `BenchmarkLikeOps` which uses a Bytes
vector. Due to the way `Bytes.Copy` is implemented, the Bytes vector's
buffer could grow arbitrarily large because the test operator never
reset the vector.

Release note: None

yuzefovich committed:
sql: propagate limit for top K sort correctly in tests

In 22.1 time frame we started propagating the value of K for top K sort
in the spec of the processor, and not in the post-processing spec, but
we forgot to update some of the tests accordingly.

Release note: None

adityamaru committed:
backupccl: fix nil pointer exception in restore OnFailOrCancel

The execCfg field on the restore resumer was not being set correctly
in the OnFailOrCancel hook. This would lead to a nil pointer exception
during cleanup.

Other jobs such as import, backup use the execCfg directly off the
JobExecContext instead of storing it on the resumer. An alternative to
this fix could be to change all `r.ExecCfg` to use the execCfg on the JobExecContext.

Fixes: #76720

Release note (bug fix): Fixes a nil pointer exception during the
cleanup of a failed or cancelled restore job.

kvoli committed:
kvserver: gossip l0sublevels instead of read amp

Previously read amplification was gossipped among stores to enable
future allocation decisions that would avoid candidates with high read
amplification. L0 sublevels represents the number of levels within the
levle 0 and is normally the significant portion of read amplification.
When read amplification is high (>15) it is normally due to a similarly
high count of L0 sublevels. This patch change read amplification to l0
sublevels, as it is a better indicator of store health.

Release justification: low risk, replace deprecated gossip signal.

Release note: None

aayushshah15 committed:
kvserver: make the StoreRebalancer interval a cluster setting

Release note (ops change): the `kv.allocator.load_based_rebalancing_interval`
cluster setting now lets operators the interval at which each store in the
cluster will check for load-based lease or replica rebalancing opportunities.

yuzefovich committed:
sql: bump --max-sql-memory in a couple of tests

The current value was small enough that we were hitting an out of memory
error during the test setup. This was caused by the fact that we now
have more memory accounting in place.

Release note: None

RichardJCai committed:
sql: support SHOW SUPER REGIONS FROM DATABASE

Release note (sql change): Add support for SHOW SUPER REGIONS FROM DATABASE.
Example:

SHOW SUPER REGIONS FROM DATABASE mr2
----
mr2  ca-central-sr  {ca-central-1}
mr2  test           {ap-southeast-2,us-east-1}

mgartner committed:
sql: deflake unique logic test

PR #78685 changes the query plan of one query in a logic test in a way
that makes the test flakey. This commit guarantees that the test cannot
be flakey, regardless of the query plan.

Release note: None

mgartner committed:
sql: ignore non-existent columns when injecting stats

Previously, an `ALTER TABLE ... INJECT STATS` command would return an
error if the given stats JSON included any columns that were not present
in the table descriptor. Statistics in statement bundles often include
dropped columns, so reproducing issues with a bundle required tediously
removing stats for these columns. This commit changes the stats
injection behavior so that a notice is issued for stats with
non-existent columns rather than an error. Any stats for existing
columns will be injected successfully.

Informs #68184

Release note (sql change): `ALTER TABLE ... INJECT STATISTICS ...` will
now issue notices when the given statistics JSON includes non-existent
columns, rather than resulting in an error. Any statistics in the JSON
for existing columns will be injected successfully.

Commit:
catalog: protobuf and cluster versioning for autostats table settings

This commit adds protobuf and cluster versioning needed for #78110

Release justification: Protobuf changes needed for autostats table settings

Release note: none

rail committed:
release: generate release archive checksums

Previously, we published release archives without their checksums.
Without checksums there was a risk of downloading corrupt files without
a way to verify their integrity.

This patch adds a new functionality to generate checksums for the
release archives and upload them next to the archive files.

Fixes #79059

Release note: None

ajwerner committed:
sql/schemachanger: rework gcjob creation for descriptions

This should also be a bit easier to extend in directions we'll need to go in
the future.

Part of dealing with #78030.

Release note: None

michae2 committed:
catalog: protobuf and cluster versioning for table statistics forecasts

Add a field to the TableDescriptor for per-table statistics forecasting.
We'll want this when backporting statistics forecasts to 22.1.

Release note: None

Release justification: low risk update to new functionality. This
creates an inert field in TableDescriptor that we'll need to use for
future backports.

ajwerner committed:
sql/parser: remove mention of BUCKET_COUNT from help text

Fixes #79044

Release note (sql change): Help text for creating indexes or primary key
constraints no longer mentions BUCKET_COUNT because it can now be omitted
and a default is used.

RichardJCai committed:
sql: Node user always passes CheckAnyPrivilege

No release note since this only affects what is shown
in InternalExecutor queries.

Release note: None

rickystewart committed:
dev,bazel: don't do `nogo` checks by default; have `doctor` advise

Up until this point `nogo` has been enabled by default and we've
required opting out with `--config nonogo`. Unfortunately `nogo` has a
non-negligible performance impact (see #73944) so having it on by
default is not what we want to be doing. Preserving the current behavior
as opt-in is still desirable though.

We have `dev doctor` advise about this case and recommend that you
explicitly set one of `--config lintonbuild` or
`--config nolintonbuild`. If you have neither or both configured,
`doctor` will detect that. (You can still override one with the other on
the command line.) We also have `dev lint` build `cockroach-short` with
`nogo` if running in non-`short` mode to cover that case for people who
have opted out of `nogo`.

`--config nonogo` is still supported as an alias for `nolintonbuild`.

Closes #73944.
Closes #78666.

Release note: None

cucaroach committed:
release-22.1: sql: fix join reader errors hitting memory limit

Restore non-streamer memory limit lower bound to 8MiB for join reader.
Streamer can handle a lower limit so we lowered it to 100KiB but that is
too low for some cases (like sqllitetests).  Also tpchvec roachtest
has 650KiB lower limit for workmem which is too low for query 20 so
restoring the 8MiB lower bound will fix that as well.

In master where streamer is on limit remains 100KiB.

Fixes: #78247
Fixes: #78341
Fixes: #77929

Release note: None

Release justification: fixes tests and restores previous behavior.

rafiss committed:
pgwire: avoid internal error for tuple in prepared statement

Neither Postgres nor CRDB can handle this, but before this change, CRDB
would show an internal error that would cause a Sentry report. Now this
is fixed.

Release note: None

erikgrinaker committed:
kv: add txn helpers that configure admission control

This patch adds the functions `DB.TxnWithAdmissionControl()` and
`kv.NewTxnWithAdmissionControl()`, which allow the caller to set the
admission control source and priority. The default variants of these
functions use source `OTHER` which bypasses admission control.

Release note: None

erikgrinaker committed:
sql/rowexec: subject column backfills to admission control

This patch subjects column backfills to admission control, using normal
priority.

Release note (bug fix): `ALTER TABLE [ADD|DROP] COLUMN` are now subject
to admission control, which will prevent these operations from
overloading the storage engine.

dt committed:
jobs: record creation cluster ID and version

Release note: none.

dt committed:
sql/importer: remove stray log line

Release note: none.

irfansharif committed:
spanconfig: ensure single reconciliation job post-restore

Fixes #70173. When restoring a cluster, we don't want to end up with two
instances of the singleton reconciliation job.

Release note: None

irfansharif committed:
backupccl: run TestFullClusterBackup with span configs

Closes #75060; speculating that the previous commit fixed the underlying
issue. Previously the reconciliation job errored out with "expected to
delete 11 row(s), deleted 10", which could conceivably occur if there
were two instances of the reconciliation job running post-restore.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79279)
<!-- Reviewable:end -->

rafiss commented:
some of the tests will need adjusting, since on the release-22.1 branch, the `enable_implicit_transaction_for_batch_statements` setting defaults to false. (but it's true on master)

fqazi committed:
sql: update tests to expect new multi-statement errors

Previously, implicit transactions could only consist of
a single statement, so error messages would always mention
implicit transactions. This is inadequate because we now
block operations implicit transactions with multiple statements
for certain operations (SET CLUSTER SETTINGS, etc..).
To address this this patch updates the expected error messages
inside test cases.

Release note: None
